File: leetcode/cpp/498.diagonal-traverse.cpp
#include "oj_header.h"

class Solution
{
public:
    vector<int> findDiagonalOrder(vector<vector<int>> &matrix)
    {
        if (matrix.empty() || matrix[0].empty())
            return {};

        vector<vector<int>> m;

        for (size_t row = 0; row < matrix.size(); row++)
        {
            for (size_t col = 0; col < matrix[row].size(); col++)
            {
                if (row + col >= m.size())
                {
                    m.push_back({});
                }
                m[row + col].push_back(matrix[row][col]);
            }
        }
        bool even = true;

        vector<int> v{};

        for (auto &line : m)
        {
            if (!even)
            {
                v.insert(end(v), begin(line), end(line));
            }
            else
            {
                v.insert(end(v), rbegin(line), rend(line));
            }

            even = !even;
        }

        return v;
    }
};

File: leetcode/kt/498.diagonal-traverse.kt
class Solution {
    fun findDiagonalOrder(matrix: Array<IntArray>): IntArray {
        if (matrix.isEmpty() || matrix[0].isEmpty()) return IntArray(0)

        val m = ArrayList<ArrayList<Int>>()
        for (i in matrix.indices) {
            for (j in matrix[i].indices) {
                if (i + j >= m.size) {
                    m.add(ArrayList())
                }

                m[i + j].add(matrix[i][j])
            }
        }

        for (i in m.indices) {
            if (i % 2 == 0) {
                m[i].reverse()
            }
        }

        return m.flatten().toIntArray()
    }
}
